### PR TITLE
If applied this commit implements a flexible social login mechanism with already active GitHub login

### PR DESCRIPTION
**Pull request by introwit**
> Better version of #211 

**Code Flow:**
- 2 flexible routes to handle all social login services.
- An array in `config/social.php` file that lists the services to be allowed to pass through the middleware.
- A middleware `app/Http/Middleware/Social.php` that checks is the incoming social login allowed in the above array or not.
- A migration `CreateUserSocialsTable` to create the new table that will consist of all the social login data.
- Covers all duplicate logins & signup scenarios.

**Steps to get this working:**

- Run `composer update` since Laravel Socialite is added.
- Run `php artisan migrate` to migrate the new required social login table.
- Update the `.env` file with GitHub credentials.

**To add & remove a new social login:**

- Update the array in `config/social.php` file to allow the respective social login to pass the middleware.
- Remove the name from the array in `config/social.php` file so the middleware will not allow it go through.

**UI**

---

PS: Ordered imported namespaces alphabetically as @mattstauffer taught me & CSS properties are also alphabetized as learned from @imjohnbon :relaxed:

### CODE REVIEW

File: app/Http/Controllers/Auth/SocialLoginController.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+use App\Http\Requests;
+use App\User;
+use Illuminate\Http\Request;
+use Socialite;
+
+class SocialLoginController extends Controller
+{
+    public function __construct()
+    {
+        $this->middleware(['social', 'guest']);
```
Comment (imjohnbon) [inline app/Http/Controllers/Auth/SocialLoginController.php]: Let's move this middleware declaration out of the controller and into the routes file (in a route group or something if it effects multiple routes).

File: app/Http/Controllers/Auth/SocialLoginController.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        return Socialite::driver($service)->redirect();
+    }
+
+    public function callback($service, Request $request)
+    {
+        $serviceUser = Socialite::driver($service)->user();
+
+        $user = $this->getExistingUser($serviceUser, $service);
+
+        if (!$user) {
```
Comment (imjohnbon) [inline app/Http/Controllers/Auth/SocialLoginController.php]: Let's add a space after the !

File: app/Http/Controllers/Auth/SocialLoginController.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            $user = User::create([
+                'name' => $serviceUser->getName(),
+                'email' => $serviceUser->getEmail(),
+            ]);
+        }
+
+        if (!$user->hasSocialLinked($service)) {
```
Comment (imjohnbon) [inline app/Http/Controllers/Auth/SocialLoginController.php]: Let's add a space after the !

File: app/Http/Middleware/Social.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if (!in_array(strtolower($request->service), array_keys(config('social.services')))) {
```
Comment (imjohnbon) [inline app/Http/Middleware/Social.php]: Let's add a space after the !

File: app/User.php [hunk 5]
```diff
@@ -119,4 +119,13 @@ protected static function boot()
             \DB::table('favorites')->where('user_id', $user->id)->delete();
         });
     }
+
+    public function social()
+    {
+        return $this->hasMany(UserSocial::class);
+    }
```
Comment (imjohnbon) [inline app/User.php]: Let's add a newline after this method.

File: resources/assets/sass/app.scss [hunk 6]
```diff
@@ -433,3 +433,141 @@ legend {
...
+
+.btn-auth:hover,
+.btn-auth:focus,
+.btn-auth:active {
+    color: #222;
+    text-decoration: none;
+}
+
+.btn-auth:before {
+    background: url(https://raw.githubusercontent.com/necolas/css3-social-signin-buttons/gh-pages/auth-icons.png) no-repeat 99px 99px;
```
Comment (imjohnbon) [inline resources/assets/sass/app.scss]: Let's move this background image to the repo instead of loading it from a third party URL.

**Comment by introwit** [comment]
@imjohnbon fixed everything else, but running `gulp` seems to make the img url go back to the old one. Did some research and this seems to be a known issue. Why these automation tools are so confusing :cry: Let me know should I push the remaining changes or the next step to make. Will research and try to solve this myself till then.

**Comment by imjohnbon** [comment]
@introwit Thanks for the update! You can feel free to push the remaining changes that you had worked on. If the image issue is still happening in Gulp, I can take a look at it later today :).

**Comment by introwit** [comment]
@imjohnbon done :)

**Comment by imjohnbon** [comment]
@introwit Thanks! I was able to make the update to load the background image locally. Not sure what was going wrong before, is it possible you were accidentally editing the compiled `app.css` instead of the `app.scss`? That's the only thing I can really think of. Anyway, this is looking good to me. I'm going to pass it along for final review. 😄

**Comment by introwit** [comment]
@imjohnbon I was editing `app.scss` only but after running `gulp` it was going back to the previous url. I know it sounds weird but that was the thing happening 😅 Thank you for approving the changes, learned new things because of this PR ❤️

File: database/migrations/2017_07_15_125034_CreateUserSocialsTable.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExists('user_socials');
```
Comment (josecanhelp) [inline database/migrations/2017_07_15_125034_CreateUserSocialsTable.php]: For the down migration, we need to 
1. drop the foreign key constraint
2. fix the misspelling of the table name

It should look like this:

```php
public function down()
    {
        Schema::table('users_social', function (Blueprint $table) {
            $table->dropForeign(['user_id']);
        });
        
        Schema::dropIfExists('users_social');
    }
```

**Pull request merged**